samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,253 @@
+"""
+InfraSyncExecutor class which runs build, package and deploy contexts
+"""
+import logging
+import re
+from typing import Dict, List, Set
+
+from boto3 import Session
+from botocore.exceptions import ClientError
+
+from samcli.commands._utils.template import get_template_data
+from samcli.commands.build.build_context import BuildContext
+from samcli.commands.deploy.deploy_context import DeployContext
+from samcli.commands.package.package_context import PackageContext
+from samcli.lib.providers.sam_stack_provider import is_local_path
+from samcli.lib.utils.resources import (
+    AWS_APIGATEWAY_RESTAPI,
+    AWS_APIGATEWAY_V2_API,
+    AWS_LAMBDA_FUNCTION,
+    AWS_LAMBDA_LAYERVERSION,
+    AWS_SERVERLESS_API,
+    AWS_SERVERLESS_FUNCTION,
+    AWS_SERVERLESS_HTTPAPI,
+    AWS_SERVERLESS_LAYERVERSION,
+    AWS_SERVERLESS_STATEMACHINE,
+    AWS_STEPFUNCTIONS_STATEMACHINE,
+    SYNCABLE_RESOURCES,
+)
+from samcli.yamlhelper import yaml_parse
+
+LOG = logging.getLogger(__name__)
+
+REMOVAL_MAP = {
+    AWS_SERVERLESS_FUNCTION: ["CodeUri", "ImageUri"],
+    AWS_LAMBDA_FUNCTION: {"Code": ["ImageUri", "S3Bucket", "S3Key", "S3ObjectVersion"]},
+    AWS_SERVERLESS_LAYERVERSION: ["ContentUri"],
+    AWS_LAMBDA_LAYERVERSION: ["Content"],
+    AWS_SERVERLESS_API: ["DefinitionBody"],
+    AWS_APIGATEWAY_RESTAPI: ["BodyS3Location"],
+    AWS_SERVERLESS_HTTPAPI: ["DefinitionUri"],
+    AWS_APIGATEWAY_V2_API: ["BodyS3Location"],
+    AWS_SERVERLESS_STATEMACHINE: ["DefinitionUri"],
+    AWS_STEPFUNCTIONS_STATEMACHINE: ["DefinitionS3Location"],
+}
```

Comment:
I would recommend keeping same types of values in a dictionary. This is hard to maintain and use.

Reply:
What do you mean by this? You mean reversing the keys and values? That makes iterations harder

Reply:
`REMOVAL_MAP` dictionary has `str` as its keys, but it has either `Dict` or `List` for its values. Having different types as the values could be confusing, so I would recommend keeping one type there.

Reply:
But this is due to lambda function having one additional layer here, under code we are storing the fields as list as well. So if I'm understanding correctly you are suggesting to change the lists into empty value dicts, which will help with the type formality, but also concerns me being not readable

Reply:
Currently changed all list notations to set notation, however it is still inconsistent in terms of typing.

Reply:
I think the best way to format this would be:
```
REMOVAL_MAP = {
    AWS_SERVERLESS_FUNCTION: {"CodeUri": [], "ImageUri": []},
    AWS_LAMBDA_FUNCTION: {"Code": ["ImageUri", "S3Bucket", "S3Key", "S3ObjectVersion"]},
    AWS_SERVERLESS_LAYERVERSION: {"ContentUri": []},
    AWS_LAMBDA_LAYERVERSION: {"Content": []},
    AWS_SERVERLESS_API: {"DefinitionBody": []},
    AWS_APIGATEWAY_RESTAPI: {"BodyS3Location": []},
    AWS_SERVERLESS_HTTPAPI: {"DefinitionUri": []},
    AWS_APIGATEWAY_V2_API: {"BodyS3Location": []},
    AWS_SERVERLESS_STATEMACHINE: {"DefinitionUri": []},
    AWS_STEPFUNCTIONS_STATEMACHINE: {"DefinitionS3Location": []},
}
```
Opinions?